maintenance/log/sink.go

```diff
@@ -0,0 +1,98 @@
+package log
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/hlog"
+	"github.com/rs/zerolog/log"
+)
+
+type sinkKey struct{}
+
+// WithSink wraps the given context in a new context with
+// the given Sink stored as value.
+func WithSink(ctx context.Context, sink *Sink) context.Context {
+	return context.WithValue(ctx, sinkKey{}, sink)
+}
+
+// SinkFromContext returns the Sink of the given context if it exits
+func SinkFromContext(ctx context.Context) (*Sink, bool) {
+	sink, ok := ctx.Value(sinkKey{}).(*Sink)
+	return sink, ok
+}
+
+// CustomOutput can be used as option func in
+// SinkWith and sets the output writer to the given one.
+func CustomOutput(w io.Writer) func(*Sink) {
+	return func(s *Sink) {
+		s.output = w
+	}
+}
+
+// Sink respresents a log sink which is used to store
+// logs, created with log.Ctx(ctx), inside the context
+// and use them at a later point in time
+type Sink struct {
+	logs []string
+
+	output  io.Writer
+	rwmutex sync.RWMutex
+}
+
+// SinkWith is a convenience function to instantiate
+// a Sink with special options like CustomOutput()
+func SinkWith(opts ...func(*Sink)) *Sink {
+	sink := &Sink{}
+
+	for _, o := range opts {
+		o(sink)
+	}
+
+	return sink
+}
+
+// HandlerWithLogSink returns a mux.MiddlewareFunc which
+// adds the given logger together with a Sink to the request
+// context. All logs corresponding to the request will be
+// printed and stored in the Sink for later use.
+func HandlerWithLogSink(logger zerolog.Logger, opts ...func(*Sink)) mux.MiddlewareFunc {
```

Comment:
I think the naming is off for most functions. We have `log.HandlerWithLogSink` that doesn't take a `Sink` argument. We have `log.CustomOutput` that's only applicable to `log.SinkWith` and `log.HandlerWithLogSink`. And it kind of looks good for `log.SinkWith(log.CustomOutput(...))`, but not for `log.HandlerWithLogSink(logger, log.CustomOutput(...))`. The naming looks like you actually wanted to do a `package sink`, but you are restricted to the `log` package.

I think using functional options right now is overkill. Actually, I think using any arguments or configuration for this is not needed for now, especially as we don't really know what we want to configure. I don't expect any real impact (performance wise, for example) from just enabling the sink in `log.Handler()` generally. Don't expose any new functions or types. We don't need any of that right now. And if we need configuration in the future, we are still able to add configuration.